man/u6a.1
.TH "U6A" "1" "Jan 30, 2020" "0.1.1" "U6A User Manual"
.
.SH NAME
u6a - Runtime for the Unlambda programming language
.
.SH SYNOPSIS
.B u6a
.RI [ options ]
.I bytecode-file
.
.SH DESCRIPTION
Load and execute Unlambda bytecode from the given
.IR bytecode-file ,
or 
.B STDIN
if "-" is given.
.
.SH OPTIONS
.TP
\fB\-s\fR, \fB\-\-stack\-segment\-size\=\fIelem-count\fR
Specify size of each stack segment of Unlambda VM to
.IR elem-count .
Default: 256.
.TP
\fB\-p\fR, \fB\-\-pool\-size\fR=\fIelem-count\fR
Specify size of object pool of Unlambda VM to
.IR elem-count .
Deafult: 1048576.
.TP
\fB\-i\fR, \fB\-\-info\fR
Print info (version, segment size, etc.) corresponding to the
.IR bytecode-file ,
then exit.
.TP
\fB\-f\fR, \fB\-\-force\fR
Attempt to execute even when the
.I bytecode-file
version is not compatible.
Meanwhile, ignore unrecognizable instructions and data during execution. 
.TP
\fB\-H\fR, \fB\-\-help\fR
Prints help message, then exit.
.TP
\fB\-V\fR, \fB\-\-version\fR
Prints version number, then exit.
.
.SH NOTES
.SS Bytecode Files
.TP
Version compatiblility:
Definition of Unlambda bytecode may differ across multiple versions of u6a.
Execution result is guaranteed to be consistent when both major and minor versions of bytecode file and the interpreter matches.
Otherwise, the code may not work as expected and the interpreter will refuse to execute unless 
.B -f
option is provided.
.TP
Redundant data:
While reading data from
.IR bytecode-file ,
any bytes before the first occurrence of magic number 
.I 0xDC
is ignored.
The same is true for bytes after 
.I .rodata
segment, however, if read from
.BR STDIN ,
they could be read by the current Unlambda program.
.
.SH SEE ALSO
.BR u6ac (1)
.
.SH COPYRIGHT
Copyright (c)  2020  CismonX <admin@cismon.net>
.PP
Copying and distribution of this file, with or without modification, are permitted in any medium without royalty, provided the copyright notice and this notice are preserved.
This file is offered as-is, without any warranty.


src/u6a.c
/*
 * u6a.c - Unlambda runtime CLI
 * 
 * Copyright (C) 2020  CismonX <admin@cismon.net>
 *
 * This file is part of U6a.
 *
 * U6a is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * U6a is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with U6a.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "logging.h"
#include "vm_defs.h"
#include "runtime.h"

#include <string.h>
#include <stdlib.h>
#include <getopt.h>
#include <errno.h>

#define EC_ERR_OPTIONS  1
#define EC_ERR_INIT     2
#define EC_ERR_RUNTIME  3

#define PARSE_UINT_OPT(opt, min_val, max_val)                            \
    errno = 0;                                                           \
    (opt) = strtoul(optarg, NULL, 10);                                   \
    if (UNLIKELY(errno)) {                                               \
        u6a_err_invalid_uint(err_toplevel, optarg);                      \
        return false;                                                    \
    }                                                                    \
    if (UNLIKELY((opt) < (min_val) || (opt) > (max_val))) {              \
        u6a_err_uint_not_in_range(err_toplevel, min_val, max_val, opt);  \
        return false;                                                    \
    }

struct arg_options {
    struct u6a_runtime_options runtime;
    bool                       print_info;
    bool                       print_only;
};

static const char* err_toplevel = "error";

static void
arg_options_destroy(struct arg_options* options) {
    if (options->runtime.istream && options->runtime.istream != stdin) {
        fclose(options->runtime.istream);
    }
}

static bool
process_options(struct arg_options* options, int argc, char** argv) {
    static const struct option long_opts[] = {
        { "stack-segment-size", required_argument, NULL, 's' },
        { "pool-size",          required_argument, NULL, 'p' },
        { "info",               no_argument,       NULL, 'i' },
        { "force",              no_argument,       NULL, 'f' },
        { "help",               no_argument,       NULL, 'H' },
        { "version",            no_argument,       NULL, 'V' },
        { 0, 0, 0, 0 }
    };
    options->runtime.stack_segment_size = U6A_VM_DEFAULT_STACK_SEGMENT_SIZE;
    options->runtime.pool_size = U6A_VM_DEFAULT_POOL_SIZE;
    options->print_info = false;
    while (true) {
        int result = getopt_long(argc, argv, "s:p:ifHV", long_opts, NULL);
        if (result == -1) {
            break;
        }
        switch (result) {
            case 's':
                PARSE_UINT_OPT(options->runtime.stack_segment_size,
                    U6A_VM_MIN_STACK_SEGMENT_SIZE, U6A_VM_MAX_STACK_SEGMENT_SIZE);
                break;
            case 'p':
                PARSE_UINT_OPT(options->runtime.pool_size, U6A_VM_MIN_POOL_SIZE, U6A_VM_MAX_POOL_SIZE);
                break;
            case 'i':
                options->print_info = true;
                break;
            case 'f':
                options->runtime.force_exec = true;
                break;
            case 'H':
                printf("Usage: u6a [options] bytecode-file\n\n"
                       "Runtime for the Unlambda programming language.\n"
                       "See \"man u6a\" for details.\n");
                options->print_only = true;
                break;
            case 'v':
                printf("%d.%d.%d\n", U6A_VER_MAJOR, U6A_VER_MINOR, U6A_VER_PATCH);
                options->print_only = true;
                break;
            case '?':
                return false;
            default:
                U6A_NOT_REACHED();
        }
    }
    if (UNLIKELY(options->print_only)) {
        return true;
    }
    if (UNLIKELY(optind == argc)) {
        u6a_err_no_input_file(err_toplevel);
        return false;
    }
    options->runtime.file_name = argv[optind];
    uint32_t file_name_size = strlen(options->runtime.file_name);
    if (file_name_size == 1 && options->runtime.file_name[0] == '-') {
        options->runtime.istream = stdin;
        options->runtime.file_name = "STDIN";
    } else if (UNLIKELY(file_name_size > PATH_MAX - 1)) {
        u6a_err_path_too_long(err_toplevel, PATH_MAX - 1, file_name_size);
        return false;
    } else {
        options->runtime.istream = fopen(options->runtime.file_name, "r");
        if (options->runtime.istream == NULL) {
            u6a_err_cannot_open_file(err_toplevel, options->runtime.file_name);
            return false;
        }
    }
    return true;
}

int main(int argc, char** argv) {
    struct arg_options options = { 0 };
    int exit_code = 0;
    u6a_logging_init(argv[0]);
    if (UNLIKELY(!process_options(&options, argc, argv))) {
        exit_code = EC_ERR_OPTIONS;
        goto terminate;
    }
    if (options.print_only) {
        goto terminate;
    }
    if (options.print_info) {
        if (UNLIKELY(!u6a_runtime_info(options.runtime.istream, options.runtime.file_name))) {
            exit_code = EC_ERR_INIT;
        }
        goto terminate;
    }
    if (UNLIKELY(!u6a_runtime_init(&options.runtime))) {
        exit_code = EC_ERR_INIT;
        goto terminate;
    }
    struct u6a_vm_var_fn exec_result = u6a_runtime_execute(stdin, stdout);
    if (UNLIKELY(U6A_VM_VAR_FN_IS_EMPTY(exec_result))) {
        exit_code = EC_ERR_RUNTIME;
        goto terminate;
    }

    terminate:
    u6a_runtime_destroy();
    arg_options_destroy(&options);
    return exit_code;
}
